src/jit/valuenum.cpp

```diff
@@ -4488,75 +4498,93 @@ void Compiler::fgValueNumberBlock(BasicBlock* blk)
         }
     }
 
-    // Now do the same for "Heap".
-    // Is there a phi for this block?
-    if (blk->bbHeapSsaPhiFunc == nullptr)
-    {
-        fgCurHeapVN = GetHeapPerSsaData(blk->bbHeapSsaNumIn)->m_vnPair.GetLiberal();
-        assert(fgCurHeapVN != ValueNumStore::NoVN);
-    }
-    else
+    // Now do the same for memory.
+    for (MemoryKind memoryKind : allMemoryKinds())
```

Comment:
We do check `byrefStatesMatchGcHeapStates`, on [line 4512](https://github.com/dotnet/coreclr/pull/9169/files#diff-dec96d75e3fa9427319bc0a7fd0213a9R4512) and [line 4572](https://github.com/dotnet/coreclr/pull/9169/files#diff-dec96d75e3fa9427319bc0a7fd0213a9R4572).  It made more sense to push them into the `else` branch of the conditional in the loop, because the `if` branch needs to do the same processing for both `memoryKind`s regardless.